File: apps/vertx-service/src/main/java/com/myapp/service/UserService.java
package com.myapp.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class UserService {
    
    private static final Logger logger = LoggerFactory.getLogger(UserService.class);
    
    // Get all users
    public static List<Map<String, Object>> getAllUsers() {
        logger.info("UserService | getAllUsers");
        
        // Dummy data for all users
        List<Map<String, Object>> users = new ArrayList<>();
        users.add(Map.of("id", 1, "name", "John Doe", "email", "john.doe@example.com"));
        users.add(Map.of("id", 2, "name", "Jane Doe", "email", "jane.doe@example.com"));
        users.add(Map.of("id", 3, "name", "Bob Smith", "email", "bob.smith@example.com"));
        
        logger.info("Returning {} users", users.size());
        return users;
    }
    
    // Get user by ID
    public static Optional<Map<String, Object>> getUserById(String userId) {
        logger.info("UserService | getUserById for ID: {}", userId);
        
        if (userId == null || userId.isEmpty()) {
            logger.warn("User ID is null or empty");
            return Optional.empty();
        }
        
        try {
            int id = Integer.parseInt(userId);
            
            // Dummy data - in real application this would come from database
            Map<String, Object> user = new HashMap<>();
            user.put("id", id);
            user.put("name", "User " + id);
            user.put("email", "user" + id + "@example.com");
            
            logger.info("User found: {}", user);
            return Optional.of(user);
            
        } catch (NumberFormatException e) {
            logger.error("Invalid user ID format: {}", userId, e);
            return Optional.empty();
        }
    }
    
    // Delete user by ID
    public static Optional<Map<String, String>> deleteUserById(String userId) {
        logger.info("UserService | deleteUserById for ID: {}", userId);
        
        if (userId == null || userId.isEmpty()) {
            logger.warn("User ID is null or empty");
            return Optional.empty();
        }
        
        try {
            int id = Integer.parseInt(userId);
            
            // Dummy delete operation - in real application this would delete from database
            // Here we just simulate successful deletion
            Map<String, String> response = new HashMap<>();
            response.put("message", "User with ID " + id + " deleted successfully");
            response.put("deletedUserId", userId);
            
            logger.info("User deleted successfully: {}", response);
            return Optional.of(response);
            
        } catch (NumberFormatException e) {
            logger.error("Invalid user ID format: {}", userId, e);
            return Optional.empty();
        }
    }
    
    // Create new user (bonus method for future use)
    public static Map<String, Object> createUser(Map<String, Object> userData) {
        logger.info("UserService | createUser with data: {}", userData);
        
        // Dummy user creation - in real application this would save to database
        Map<String, Object> newUser = new HashMap<>(userData);
        newUser.put("id", (int) (Math.random() * 1000)); // Generate random ID
        newUser.put("createdAt", System.currentTimeMillis());
        
        logger.info("User created successfully: {}", newUser);
        return newUser;
    }
    
    // Update user (bonus method for future use)
    public static Optional<Map<String, Object>> updateUser(String userId, Map<String, Object> userData) {
        logger.info("UserService | updateUser for ID: {} with data: {}", userId, userData);
        
        if (userId == null || userId.isEmpty()) {
            logger.warn("User ID is null or empty");
            return Optional.empty();
        }
        
        try {
            int id = Integer.parseInt(userId);
            
            // Dummy update operation - in real application this would update in database
            Map<String, Object> updatedUser = new HashMap<>(userData);
            updatedUser.put("id", id);
            updatedUser.put("updatedAt", System.currentTimeMillis());
            
            logger.info("User updated successfully: {}", updatedUser);
            return Optional.of(updatedUser);
            
        } catch (NumberFormatException e) {
            logger.error("Invalid user ID format: {}", userId, e);
            return Optional.empty();
        }
    }
}

File: apps/actix-service/src/services/user_service.rs
use serde_json::json;
use crate::helper::logger;

pub struct UserService;

impl UserService {
    pub fn new() -> Self {
        UserService
    }

    pub async fn get_users(&self) -> Result<Vec<serde_json::Value>, String> {
        logger::info("Starting to fetch users from service layer").await;

        let users = vec![
            json!({ "id": 1, "name": "Alice", "email": "alice@example.com" }),
            json!({ "id": 2, "name": "Bob", "email": "bob@example.com" }),
            json!({ "id": 3, "name": "Charlie", "email": "charlie@example.com" })
        ];

        logger::info(format!("Successfully retrieved {} users from service layer", users.len())).await;
        
        Ok(users)
    }

    pub async fn get_user_by_id(&self, user_id: u32) -> Result<Option<serde_json::Value>, String> {
        logger::info(format!("Fetching user with ID: {}", user_id)).await;

        let users = self.get_users().await?;
        let user = users.into_iter().find(|u| u["id"].as_u64() == Some(user_id as u64));

        match user {
            Some(found_user) => {
                logger::info(format!("Successfully found user with ID: {}", user_id)).await;
                Ok(Some(found_user))
            }
            None => {
                logger::warn(format!("User with ID {} not found", user_id)).await;
                Ok(None)
            }
        }
    }

    pub async fn create_user(&self, user_data: serde_json::Value) -> Result<serde_json::Value, String> {
        logger::info("Creating new user in service layer").await;

        if !user_data["name"].is_string() || !user_data["email"].is_string() {
            return Err("Invalid user data: name and email are required".to_string());
        }

        let new_user = json!({
            "id": 4,
            "name": user_data["name"],
            "email": user_data["email"]
        });

        logger::info("Successfully created new user in service layer").await;
        Ok(new_user)
    }

    pub async fn update_user(&self, user_id: u32, user_data: serde_json::Value) -> Result<Option<serde_json::Value>, String> {
        logger::info(format!("Updating user with ID: {}", user_id)).await;

        let existing_user = self.get_user_by_id(user_id).await?;
        if existing_user.is_none() {
            return Ok(None);
        }

        let updated_user = json!({
            "id": user_id,
            "name": user_data["name"].as_str().unwrap_or("Unknown"),
            "email": user_data["email"].as_str().unwrap_or("unknown@example.com"),
            "updated_at": chrono::Utc::now().to_rfc3339()
        });

        logger::info(format!("Successfully updated user with ID: {}", user_id)).await;
        Ok(Some(updated_user))
    }

    pub async fn delete_user(&self, user_id: u32) -> Result<bool, String> {
        logger::info(format!("Deleting user with ID: {}", user_id)).await;

        let existing_user = self.get_user_by_id(user_id).await?;
        if existing_user.is_none() {
            logger::warn(format!("Cannot delete user with ID {}: not found", user_id)).await;
            return Ok(false);
        }

        logger::info(format!("Successfully deleted user with ID: {}", user_id)).await;
        Ok(true)
    }
}